core/src/main/scala/kafka/log/TransactionIndex.scala

```diff
@@ -0,0 +1,197 @@
+
+          val abortedTxn = new AbortedTxn(buffer)
+          if (abortedTxn.version > AbortedTxn.CurrentVersion)
+            trace(s"Reading aborted txn entry with version ${abortedTxn.version} as current " +
+              s"version ${AbortedTxn.CurrentVersion}")
+
+          if (abortedTxn.lastOffset >= offset) {
+            channel.truncate(position)
+            return
+          }
+          position += AbortedTxn.TotalSize
+        }
+      }
+    }
+  }
+
+  def collectAbortedTxns(fetchOffset: Long, upperBoundOffset: Long): TxnIndexSearchResult = {
+    maybeChannel match {
+      case None => TxnIndexSearchResult(abortedTransactions = List.empty[AbortedTransaction], isComplete = false)
+      case Some(channel) =>
+        val abortedTransactions = ListBuffer[AbortedTransaction]()
+        var position = 0
+        val lastPosition = channel.position()
+
+        while (lastPosition - position >= AbortedTxn.TotalSize) {
+          val buffer = ByteBuffer.allocate(AbortedTxn.TotalSize)
+          Utils.readFully(channel, buffer, position)
+          buffer.flip()
+
+          // FIXME: IF we want to do this, we need to be able to reliably determine the size across versions
+          val abortedTxn = new AbortedTxn(buffer)
+          if (abortedTxn.version > AbortedTxn.CurrentVersion)
+            trace(s"Reading aborted txn entry with version ${abortedTxn.version} as current " +
+              s"version ${AbortedTxn.CurrentVersion}")
+
+          if (abortedTxn.lastOffset >= fetchOffset && abortedTxn.firstOffset < upperBoundOffset)
+            abortedTransactions += abortedTxn.asAbortedTransaction
+
+          if (abortedTxn.lastStableOffset >= upperBoundOffset)
+            return TxnIndexSearchResult(abortedTransactions = abortedTransactions.toList, isComplete = true)
+
+          position += AbortedTxn.TotalSize
+        }
+
+        TxnIndexSearchResult(abortedTransactions = abortedTransactions.toList, isComplete = false)
+    }
+  }
+}
+
+private[log] object AbortedTxn {
+  val VersionOffset = 0
+  val VersionSize = 2
+  val ProducerIdOffset = VersionOffset + VersionSize
+  val ProducerIdSize = 8
+  val FirstOffsetOffset = ProducerIdOffset + ProducerIdSize
+  val FirstOffsetSize = 8
+  val LastOffsetOffset = FirstOffsetOffset + FirstOffsetSize
+  val LastOffsetSize = 8
+  val LastStableOffsetOffset = LastOffsetOffset + LastOffsetSize
+  val LastStableOffsetSize = 8
+  val TotalSize = LastStableOffsetOffset + LastStableOffsetSize
+
+  val CurrentVersion: Short = 0
+}
+
+private[log] class AbortedTxn(val buffer: ByteBuffer) {
+  import AbortedTxn._
+
+  def this(producerId: Long,
+           firstOffset: Long,
+           lastOffset: Long,
+           lastStableOffset: Long) = {
+    this(ByteBuffer.allocate(AbortedTxn.TotalSize))
+    buffer.putShort(CurrentVersion)
+    buffer.putLong(producerId)
+    buffer.putLong(firstOffset)
+    buffer.putLong(lastOffset)
+    buffer.putLong(lastStableOffset)
+    buffer.flip()
```

Comment:
Kind of a subjective call, but since we scan the abort index when handling fetches and the schema is so simple, I thought we could skip the need to go through the `Struct` object when reading and writing.